tekton/utils/pipeline_run_builder.go

```diff
@@ -214,29 +223,74 @@ func (b *PipelineRunBuilder) WithParamsFromConfigMap(configMap *corev1.ConfigMap
 // WithPipelineRef sets the PipelineRef for the PipelineRun's spec.
 func (b *PipelineRunBuilder) WithPipelineRef(pipelineRef *tektonv1.PipelineRef) *PipelineRunBuilder {
 	b.pipelineRun.Spec.PipelineRef = pipelineRef
+	return b
+}
 
-	if pipelineRef.Resolver == "git" {
-		for _, param := range pipelineRef.Params {
-			if param.Name == "revision" {
-				b.WithParams(tektonv1.Param{
-					Name: "taskGitRevision",
-					Value: tektonv1.ParamValue{
-						Type:      tektonv1.ParamTypeString,
-						StringVal: param.Value.StringVal,
-					},
-				})
-			}
-
-			if param.Name == "url" {
-				b.WithParams(tektonv1.Param{
-					Name: "taskGitUrl",
-					Value: tektonv1.ParamValue{
-						Type:      tektonv1.ParamTypeString,
-						StringVal: param.Value.StringVal,
-					},
-				})
-			}
+// WithGitResolution resolves git branch references to commit SHAs and adds git parameters.
+func (b *PipelineRunBuilder) WithGitResolution() *PipelineRunBuilder {
+	pipelineRef := b.pipelineRun.Spec.PipelineRef
+	if pipelineRef == nil || pipelineRef.ResolverRef.Resolver != "git" {
+		return b
+	}
+
+	var gitURL, revision string
+	var revisionIndex int = -1
+
+	for i, param := range pipelineRef.ResolverRef.Params {
+		switch param.Name {
+		case "url":
+			gitURL = param.Value.StringVal
+		case "revision":
+			revision = param.Value.StringVal
+			revisionIndex = i
+		}
+	}
+
+	addGitParams := func(finalRevision string) {
+		b.WithParams(
+			tektonv1.Param{Name: "taskGitUrl", Value: tektonv1.ParamValue{Type: tektonv1.ParamTypeString, StringVal: gitURL}},
+			tektonv1.Param{Name: "taskGitRevision", Value: tektonv1.ParamValue{Type: tektonv1.ParamTypeString, StringVal: finalRevision}},
+			tektonv1.Param{Name: "pipelineGitRevision", Value: tektonv1.ParamValue{Type: tektonv1.ParamTypeString, StringVal: finalRevision}},
+		)
+	}
+
+	resolvedSHA, err := git.ResolveBranchToSHA(gitURL, revision)
+
+	if err != nil {
+		if b.logger != nil {
```

Comment:
Usually we shouldn't need to check if the logger is nil, since we'll be sure to pass it down in `NewPipelineRunBuilder`.